src/asm/lexer.c

```diff
@@ -1125,39 +1126,63 @@ static uint32_t readNumber(int radix, uint32_t baseValue)
 	return value;
 }
 
-static uint32_t readFractionalPart(int32_t integer)
+static uint32_t readFractionalPart(uint32_t integer)
 {
 	uint32_t value = 0, divisor = 1;
+	uint8_t precision = 0;
+	enum {
+		READFRACTIONALPART_DIGITS,
+		READFRACTIONALPART_PRECISION,
+		READFRACTIONALPART_PRECISION_DIGITS
+	} state = READFRACTIONALPART_DIGITS;
 
 	for (;; shiftChar()) {
 		int c = peek();
 
-		if (c == '_')
-			continue;
-		else if (c < '0' || c > '9')
-			break;
-		if (divisor > (UINT32_MAX - (c - '0')) / 10) {
-			warning(WARNING_LARGE_CONSTANT,
-				"Precision of fixed-point constant is too large\n");
-			// Discard any additional digits
-			shiftChar();
-			while (c = peek(), (c >= '0' && c <= '9') || c == '_')
+		if (state == READFRACTIONALPART_DIGITS) {
+			if (c == '_') {
+				continue;
+			} else if (c == 'q' || c == 'Q') {
+				state = READFRACTIONALPART_PRECISION;
+				continue;
+			} else if (c < '0' || c > '9') {
+				break;
+			}
+			if (divisor > (UINT32_MAX - (c - '0')) / 10) {
+				warning(WARNING_LARGE_CONSTANT,
+					"Precision of fixed-point constant is too large\n");
+				// Discard any additional digits
 				shiftChar();
-			break;
+				while (c = peek(), (c >= '0' && c <= '9') || c == '_')
+					shiftChar();
+				break;
+			}
+			value = value * 10 + (c - '0');
+			divisor *= 10;
+		} else {
+			if (c == '.' && state == READFRACTIONALPART_PRECISION) {
+				state = READFRACTIONALPART_PRECISION_DIGITS;
+				continue;
+			} else if (c < '0' || c > '9') {
+				break;
+			}
+			precision = precision * 10 + (c - '0');
```

Comment:
Careful that this is not bounded (overflow; if > 32 = the `1 <<` below will be UB).